File: haskell/src/Day01.hs
module Day01 (solutionPartOne, solutionPartTwo) where

import Data.List (sort)

solutionPartOne :: String -> String
solutionPartOne = 
    show. 
    maximum. 
    splitElvesCalories

solutionPartTwo :: String -> String
solutionPartTwo = 
    show. 
    sumTopThree.
    splitElvesCalories

sumTopThree :: [Int] -> Int
sumTopThree = sum . take 3 . reverse . sort

sumStrings :: [String] -> Int
sumStrings = sum . map read

splitElvesCalories :: String -> [Int]
splitElvesCalories rawCalories =
    buildElvesCalories $ lines rawCalories
    where
        buildElvesCalories :: [String] -> [Int]
        buildElvesCalories [] = []
        buildElvesCalories calories = 
            let (elveCalories, othersCalories) = break (== "") calories
            in sumStrings elveCalories : buildElvesCalories (drop 1 othersCalories)

File: c#/Lib/Day01.cs
namespace Lib;

using System.Linq;

public class Day01
{
    public static string SolutionPartOne(string rawCalories)
        => SumCaloriesOfEachElves(rawCalories)
            .Max()
            .ToString();

    public static string SolutionPartTwo(string rawCalories)
        => TakeHighestThree(SumCaloriesOfEachElves(rawCalories))
            .Sum()
            .ToString();

    public static IEnumerable<int> TakeHighestThree(IEnumerable<int> elvesCalories)
        => elvesCalories.OrderByDescending(calories => calories).Take(3);

    public static IEnumerable<string> SplitElvesCalories(string @string)
        => @string.Split("\n\n");

    public static int SumCalories(string rawCalories)
        => rawCalories.Split("\n").Select(int.Parse).Sum();
    
    public static IEnumerable<int> SumCaloriesOfEachElves(string rawCaloriesOfEachElves) 
        => SplitElvesCalories(rawCaloriesOfEachElves).Select(SumCalories);
    
}